packages/attune/src/bin/attune/cmd/apt/pkg/add.rs

```diff
@@ -24,96 +26,172 @@ use attune::{
+            Err(other) => {
+                eprintln!("Unable to add package to index: {other:#?}");
+                ExitCode::FAILURE
+            }
+        },
+    }
+}
+
+/// Ensure that the specified repository exists.
+#[instrument]
+pub async fn validate_repository_exists(ctx: &Config, cmd: &PkgAddCommand) -> Result<bool> {
+    debug!("checking whether repository exists");
     let res = ctx
         .client
         .get(
             ctx.endpoint
                 .join(
                     format!(
                         "/api/v0/repositories/{}",
-                        percent_encode(command.repo.as_bytes(), PATH_SEGMENT_PERCENT_ENCODE_SET)
+                        percent_encode(cmd.repo.as_bytes(), PATH_SEGMENT_PERCENT_ENCODE_SET)
                     )
                     .as_str(),
                 )
                 .unwrap(),
         )
         .send()
         .await
-        .expect("Could not send API request");
+        .context("send api request")?;
     match res.status() {
         StatusCode::OK => {
             let repo = res
                 .json::<RepositoryInfoResponse>()
                 .await
-                .expect("Could not parse response");
+                .context("parse response")?;
             debug!(?repo, "repository exists");
+            Ok(true)
         }
         StatusCode::NOT_FOUND => {
-            eprintln!("Error: repository {:?} does not exist", command.repo);
-            return ExitCode::FAILURE;
+            debug!("repository does not exist");
+            Ok(false)
         }
         status => {
-            let body = res.text().await.expect("Could not read response");
+            let body = res.text().await.context("read response")?;
             debug!(?body, ?status, "error response");
-            let error = serde_json::from_str::<ErrorResponse>(&body)
-                .expect("Could not parse error response");
-            eprintln!(
-                "Error checking whether repository exists: {}",
-                error.message
-            );
-            return ExitCode::FAILURE;
+            let error =
+                serde_json::from_str::<ErrorResponse>(&body).context("parse error response")?;
+            bail!(error,);
         }
     }
+}
+
+/// Checksum the package file, and upload if needed.
+// TODO: We might want to make this streaming for sufficiently large package
+// files (ones that don't fit in memory). For small ones, I think keeping
+// the file in memory might be faster.
+//
+// TODO: We may also want to check whether a package with the same
+// identifiers (i.e. (name, version, architecture)) already exists, which
+// should be impossible, and should be an error we report to the user.
+//
+// TODO(#48): Add an `--overwrite` flag to allow the user to deliberately upload
+// a package with a different SHA256sum.
+#[instrument(fields(sha256sum))]
+pub async fn upsert_file_content(ctx: &Config, cmd: &PkgAddCommand) -> Result<String> {
```

Comment:
This isn't actually an upsert. In fact, the correct behavior if you upload a package with different contents but the same identifiers is an error (see #130).
```suggestion
pub async fn upload_file_content(ctx: &Config, cmd: &PkgAddCommand) -> Result<String> {
```